httpio_async/__init__.py

```diff
@@ -0,0 +1,293 @@
+            self._session = None
+            self._cache.clear()
+            self.closed = True
+
+    async def flush(self):
+        self._cache.clear()
+
+    async def peek(self, size):
+        return await self._read_impl(size, peek=True)
+
+    async def read(self, size=-1):
+        return await self._read_impl(size)
+
+    async def read1(self, size=-1):
+        return await self._read_impl(size, 1)
+
+    async def readable(self):
+        return True
+
+    async def readinto(self, b):
+        return await self._readinto_impl(b)
+
+    async def readinto1(self, b):
+        return await self._readinto_impl(b, 1)
+
+    async def readline(self, size=-1):
+        self._assert_open()
+
+        if size < 1 or self._cursor + size > self.length:
+            size = self.length - self._cursor
+
+        if size == 0:
+            return b""
+
+        if self.block_size <= 0:
+            data = await self._read_raw(self._cursor, self._cursor + size)
+
+        else:
+            data = b''
+            while len(data) < size:
+                async for sector in self._read_cached(size,
+                                                      max_raw_reads=1):
+                    data += sector
+                    if b'\n' in sector:
+                        break
+                if b'\n' in sector:
+                        break
+
+        data = data.splitlines(True)[0]
+        self._cursor += len(data)
+        return data
+
+    async def readlines(self, size=-1):
+        self._assert_open()
+
+        if size < 1 or self._cursor + size > self.length:
+            size = self.length - self._cursor
+
+        if size == 0:
+            yield b""
+        else:
+            if self.block_size <= 0:
+                data = await self._read_raw(self._cursor, self._cursor + size)
+                for line in data.splitlines(True):
+                    yield line
+            else:
+                data = b''
+                while len(data) < size:
+                    async for sector in self._read_cached(size,
+                                                          max_raw_reads=1):
+                        for line in sector.splitlines(True):
+                            data += line
+                            if data.endswith(b'\n'):
+                                yield data
+                                size -= len(data)
+                                self._cursor += len(data)
+                                data = b''
+                self._cursor += len(data)
+                yield data
+
+    async def seek(self, offset, whence=0):
+        self._assert_open()
+        if whence == 0:
+            self._cursor = offset
+        elif whence == 1:
+            self._cursor += offset
+        elif whence == 2:
+            self._cursor = self.length + offset
+        else:
+            raise HTTPIOError("Invalid argument: whence=%r" % whence)
+        if not (0 <= self._cursor <= self.length):
```

Comment:
Not related to this PR but possibly interesting to note that a seek beyond the limits should be allowed without raising an exception. A file object from `open` allows it similar to C. I noticed it only because `_read_impl` has a `if size == 0` check, not <= 0.